qemu/tests/hugepage_specify_node.py

```diff
@@ -0,0 +1,84 @@
+    Steps:
+    1) Setup total of 4G mem hugepages for specify node.
+    2) Setup total of 1G mem hugepages for idle node.
+    3) Mount this hugepage to /mnt/kvm_hugepage.
+    4) Boot guest only allocate hugepage from specify node.
+    5) Check the hugepage used from every node.
+    :params test: QEMU test object.
+    :params params: Dictionary with the test parameters.
+    :params env: Dictionary with test environment.
+    """
+    hugepage_size = memory.get_huge_page_size()
+    mem_size = int(normalize_data_size("%sM" % params["mem"], "K"))
+    idle_node_mem = int(normalize_data_size("%sM" % params["idle_node_mem"], "K"))
+
+    error_context.context("Get host numa topological structure.", logging.info)
+    host_numa_node = utils_misc.NumaInfo()
+    node_list = host_numa_node.online_nodes_withmem
+    idle_node_list = node_list.copy()
+    pre_hugepages_free = {}
+    for node_id in node_list:
+        error_context.base_context("Check preprocess HugePages Free on host "
+                                   "numa node %s." % node_id, logging.info)
+        node_memfree = int(host_numa_node.read_from_node_meminfo(node_id,
+                                                                 "MemFree"))
+        value = int(host_numa_node.read_from_node_meminfo(node_id,
+                                                          "HugePages_Free"))
+        pre_hugepages_free[node_id] = value
+        if node_memfree < idle_node_mem:
+            idle_node_list.remove(node_id)
+        if node_memfree < mem_size:
+            node_list.remove(node_id)
+
+    if len(node_list) < 2:
+        test.cancel("Host node does not have enough free memory to run the test, "
+                    "skipping test...")
+
+    for node_id in node_list:
+        error_context.base_context("Specify qemu process only allocate "
+                                   "HugePages from node%s." % node_id, logging.info)
+        params["target_nodes"] = "%s" % node_id
+        params["target_num_node%s" % node_id] = (mem_size // hugepage_size)
+        error_context.context("Setup huge pages for specify node%s." %
+                              node_id, logging.info)
+
+        for idle_node in [_ for _ in idle_node_list if _ != node_id]:
+            params["target_nodes"] += " %s" % idle_node
+            params["target_num_node%s" % idle_node] = (idle_node_mem // hugepage_size)
+            error_context.context("Setup huge pages for idle node%s." %
+                                  idle_node, logging.info)
+        params["setup_hugepages"] = "yes"
+        env_process.preprocess(test, params, env)
+        params["qemu_command_prefix"] = "numactl --membind=%s" % node_id
+        params["start_vm"] = "yes"
+        env_process.preprocess_vm(test, params, env, params["main_vm"])
+        try:
+            vm = env.get_vm(params["main_vm"])
+            vm.verify_alive()
+            vm.wait_for_login()
+
+            for index in idle_node_list:
```

Comment:
Why not use the same code in L60, so you don't need to add `index != node_id` in L80.

Reply:
Because I think it is clearer to show all the judgment conditions in L80.

Reply:
Of course, both methods are ok for me, if you want to consistency, I can modify accordingly.

Reply:
Not sure which one is better. But I guess you could use a new list to store `[_ for _ in idle_node_list if _ != node_id]`, then you don't have to exclude node_id twice.

Reply:
Thanks for the quick reply.
correct.